soci/soci_index.go

```diff
@@ -395,8 +395,13 @@ func (b *IndexBuilder) buildSociLayer(ctx context.Context, desc ocispec.Descript
 	if err != nil {
 		return nil, fmt.Errorf("could not determine layer compression: %w", err)
 	}
-	if compressionAlgo != compression.Gzip {
-		return nil, fmt.Errorf("layer %s (%s) must be compressed by gzip, but got %q: %w",
+	// `images.DiffCompression` returns "" if it's a uncompressed layer (tar).
+	if compressionAlgo == "" || compressionAlgo == "unknown" {
+		compressionAlgo = compression.Uncompressed
+	}
```

Comment:
I think we should have a separate "unknown" compression algo witch we treat as uncompressed when creating a zinfo. Eventually, we may want to test the actual file to see what compression it is to resolve this, but if we collapse it to uncompressed here then it will be harder to do that later. 

"unknown" is distinct from "unrecognized" -> the latter comes back as an error.

Reply:
Sorry I didn't quite understand the suggestion here.

I think we want to recognize the following media types as uncompresed layer:

```golang
// docker:
MediaTypeDockerSchema2Layer            = "application/vnd.docker.image.rootfs.diff.tar"
MediaTypeDockerSchema2LayerForeign     = "application/vnd.docker.image.rootfs.foreign.diff.tar"
// containerd/oci
MediaTypeImageLayer = "application/vnd.oci.image.layer.v1.tar"
MediaTypeImageLayerNonDistributable = "application/vnd.oci.image.layer.nondistributable.v1.tar"
```

For the first 2, `images.DiffCompression` returns `unknown`: https://github.com/containerd/containerd/blob/a7ceac8b63575b08e65f74c852f7dd8b3407ad64/images/mediatypes.go#LL77C24-L77C24

For the last 2, it returns `""`: https://github.com/containerd/containerd/blob/a7ceac8b63575b08e65f74c852f7dd8b3407ad64/images/mediatypes.go#LL93C10-L93C10

So in these 2 cases they both map to `uncompressed` ztoc.

Maybe we can change this to a switch similar to the `DiffCompression` func, e.g.,

```golang
switch mediaType {
case MediaTypeDockerSchema2Layer, MediaTypeDockerSchema2LayerForeign:
    if compressAlgo == "unknown" {
        compressAlgo = "uncompressed"
    }
case MediaTypeImageLayer, MediaTypeImageLayerNonDistributable:
    if compressAlgo == "" {
        compressAlgo = "uncompressed"
    }
}
```

WDYT?

Reply:
I think we should keep our options open here. Containerd does auto-detection of the compression algorithm for "unkown" so I think we should try to do the same. For now, we can skip the auto detection and just use no compression, but we should revisit that later.

```
"gzip" => compression.Gzip
"uncompressed" => compression.Uncompressed
"unknown" => compression.Unknown
```

then when creating a zinfo:
```
case Gzip:
	return newGzipZinfo(zinfoBytes)
case Uncompressed:
case Unknown:
	return newTarZinfo(zinfoBytes)
```

Reply:
update unknown handling in a separete commit. PTAL @Kern--